docs: Add documentation for contributing additional checks to the trivy policies repo (#6234)
aquasecurity/trivy: docs/community/contribute/checks/service-support.md

# Add Service Support

A service refers to a service by a cloud provider. This section details how to add a new service to an existing provider. All contributions need to be made to the [trivy repository](https://github.com/aquasecurity/trivy/).

Before you begin, verify that the provider does not already have the service that you plan to add.

> Adding a new service involves two steps. The service will need a data structure to store information about the required resources, and then one or more adapters to convert input(s) into the aforementioned data structure.

Reviewer: what are those input(s)?

## Adding a new service to an existing provider

To add a new service named `Bar` to a provider named `foo`, you'll need to add a new file at `pkg/providers/foo/bar/bar.go`:

```
type Bar struct {
    // ...
}
```

Let's say the `Bar` service manages resources called `Baz`. You'll need to add a new struct to the `Bar` struct to hold information about this resource:

```
type Bar struct {
    // ...
    Baz []Baz
    // ...
}

type Baz struct {
    types.Metadata
	Name types.StringValue
	Encrypted types.BoolValue

> }

Reviewer: indentation

Author: ah, it looked different on VS Code, not showing the indentation

```

> A `Baz` can have a name, and can optionally be encrypted. Instead of using raw string and bool types respectively, we use the trivy types `types.StringValue` and `types.BoolValue`. These types wrap the raw values and provide additional metadata about the value, such as whether it was set by the user or not, and the file and line number where the resource was defined. The `types.Metadata` struct is embedded in all of the Trivy types, and provides a common set of metadata for all resources. This includes the file and line number where the resource was defined, and the name of the resource.

Reviewer: again, need to draw the line that we are discussing an example

Reviewer: sentence about metadata can come first, as we're now covering the code and it comes first there

Next you'll need to add a reference to your new service struct in the provider struct at `pkg/providers/foo/foo.go`:

```
type Foo struct {
    // ...
    Bar bar.Bar
    // ...
}
```

> Now you'll need to update all of the adapters which populate the Foo provider struct. For example, if you want to support Terraform, you'll need to update `internal/adapters/terraform/foo/bar/adapt.go`.

Reviewer: this is over simplification.. not sure it's helpful. what I recommend is creating this doc as an example for how to add a service to AWS, and then mention the adapters needed to update for that scenario.

> Finally, make sure you run make schema to generate the schema for your new service.

Reviewer: explain what, who

Have a look at the other providers and services in the [`iac/providers`](https://github.com/aquasecurity/trivy/tree/main/pkg/iac/providers) directory in Trivy.

Next you'll need to add a reference to your new service struct in the [provider struct](https://github.com/aquasecurity/trivy/blob/main/pkg/iac/providers/aws/aws.go) at `pkg/iac/providers/aws/aws.go`:

```
type AWS struct {
	...
	CodeBuild      codebuild.CodeBuild
    ...
}
```

Now you'll need to update all of the [adapters](https://github.com/aquasecurity/trivy/tree/main/pkg/iac/adapters) which populate the AWS provider struct. For example, if you want to support Terraform, you'll need to update [`trivy/pkg/iac/adapters/terraform/aws/codebuild/adapt.go`](https://github.com/aquasecurity/trivy/blob/main/pkg/iac/adapters/terraform/aws/codebuild/adapt.go).

> TODO HOW?

Author: @simar7 I need some additional information on how people can add adapters -- i.e. what sections are required, what information should I highlight? the adapt.go file looks fairly complex

Reviewer 2: The comments on [this](https://github.com/aquasecurity/defsec/pull/1000/files) PR should help.

As for learning about the resources themselves, I would say terraform documentation is a good place to go to. For instance [this](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/db_instance.html) is an example.

## Create a new Schema for your provider

Once the new service has been added to the provider, you need to create a the schema for the service as part of the provider schema. In the Trivy root directory run `make schema` to generate the schema for your new service.

> TODO

Reviewer 2: We've updated to use `magefiles` as the rest of the Trivy targets. You can find them [here](https://github.com/aquasecurity/trivy/blob/main/magefiles/magefile.go#L429-L437)

So to generate the new schema it would be `mage schema:generate` and to verify it, the command will be `mage schema:verify`

You can see these targets by running `mage -l` in the Trivy repo.

First, create a new directory and file for your new service under the provider directory: e.g. [aws/codebuild/codebuild.go](https://github.com/aquasecurity/trivy/blob/main/pkg/iac/providers/aws/codebuild/codebuild.go)

The CodeBuild service will require a structure `struct` to hold the information on the input that is scanned. The input is the CodeBuild resource that a user configured and wants to scan for misconfiguration.

```
type CodeBuild struct {
	Projects []Project
}
```

The CodeBuild service manages `Project` resources. The `Project` struct has been added to hold information about each Project resources; `Project` Resources in turn manage `ArtifactSettings`:

```
type Project struct {
	Metadata                  iacTypes.Metadata
	ArtifactSettings          ArtifactSettings
	SecondaryArtifactSettings []ArtifactSettings
}

type ArtifactSettings struct {
	Metadata          iacTypes.Metadata
	EncryptionEnabled iacTypes.BoolValue
}
```

The `iacTypes.Metadata` struct is embedded in all of the Trivy types and provides a common set of metadata for all resources. This includes the file and line number where the resource was defined and the name of the resource.

> A resource, in this example `Project`, can have a name, and can optionally be encrypted. Instead of using raw string and bool types respectively, we use the trivy types `iacTypes.Metadata` and `iacTypes.BoolValue`. These types wrap the raw values and provide additional metadata about the value, such as whether it was set by the user or not, and the file and line number where the resource was defined. 

Reviewer 2: comma splice 

```suggestion
A resource in this example `Project` can have a name and can optionally be encrypted. Instead of using raw string and bool types respectively, we use the trivy types `iacTypes.Metadata` and `iacTypes.BoolValue`. These types wrap the raw values and provide additional metadata about the value. For instance, whether it was set by the user and the file and line number where the resource was defined. 
```
